<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Computing a PGS using RápidoPGS-single and GWAS catalog • RápidoPGS</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Computing a PGS using RápidoPGS-single and GWAS catalog">
<meta property="og:description" content="RapidoPGS">
<meta property="og:image" content="http://GRealesM.github.io/RapidoPGS/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RápidoPGS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Computing_RapidoPGSmulti.html">Computing a PGS using RápidoPGS-multi</a>
    </li>
    <li>
      <a href="../articles/Computing_RapidoPGSsingle.html">Computing a PGS using RápidoPGS-single and GWAS catalog</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/GRealesM/RapidoPGS" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Computing a PGS using RápidoPGS-single and GWAS catalog</h1>
                        <h4 data-toc-skip class="author">Guillermo Reales</h4>
            
            <h4 data-toc-skip class="date">2023-10-12</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/GRealesM/RapidoPGSvignettes/Computing_RapidoPGSsingle.Rmd" class="external-link"><code>vignettes/Computing_RapidoPGSsingle.Rmd</code></a></small>
      <div class="hidden name"><code>Computing_RapidoPGSsingle.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>RápidoPGS is a tool to quickly compute polygenic scores (PGS) from GWAS summary statistics datasets from either case-control (eg. asthma) or quantitative traits (eg. height and BMI).</p>
<p>Input GWAS summary statistics datasets should have a minimum set of columns, with these names, but in any order:</p>
<ul>
<li>
<strong>CHR</strong>: Chromosome</li>
<li>
<strong>BP</strong>: Base position (in GRCh37/hg19 or GRCh38/hg38). If using hg38, use build = “hg38” in parameters.</li>
<li>
<strong>SNPID</strong>: rsids, or SNP identifiers. If not available, they can be anything (eg. CHR_BP).</li>
<li>
<strong>REF</strong>: Reference, or non-effect allele</li>
<li>
<strong>ALT</strong>: Alternative, or effect allele, the one  refers to</li>
<li>
<strong>BETA</strong>: <span class="math inline">\(\beta\)</span> (or log(OR)), or effect sizes.</li>
<li>
<strong>SE</strong>: Standard error of <span class="math inline">\(\beta\)</span>
</li>
<li>
<strong>P</strong>: P-values for the association test</li>
</ul>
<p>Also, if doing a PGS on a quantitative trait GWAS dataset, your file must include this:</p>
<ul>
<li>
<strong>ALT_FREQ</strong>: Minor/ALT allele frequency in the tested population, or in a close population from a reference panel.</li>
</ul>
</div>
<div class="section level2">
<h2 id="installation-of-rápidopgs">Installation of RápidoPGS<a class="anchor" aria-label="anchor" href="#installation-of-r%C3%A1pidopgs"></a>
</h2>
<p>Current RápidoPGS version (v2.1.0) is available on CRAN, so we can install it using the following code</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"RapidoPGS"</span><span class="op">)</span></span></code></pre></div>
<p>Alternatively, you can download the development version from Github (Note: If you don’t have <code>remotes</code> installed, please install it first:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"remotes"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"GRealesM/RapidoPGS"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<p>Once installed, let’s load it. This will automatically load all required dependencies.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RapidoPGS</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="downloading-data-from-gwas-catalog">Downloading data from GWAS catalog<a class="anchor" aria-label="anchor" href="#downloading-data-from-gwas-catalog"></a>
</h2>
<p><a href="https://www.ebi.ac.uk/gwas/" class="external-link">GWAS catalog</a> is an outstanding GWAS summary statistics data source, as it not only puts together tons of public datasets, but it also uses a semi-automatic pipeline to apply quality control and liftover (a.k.a. harmonise) those datasets, thus helping overcome the problem of having GWAS sumstats in so many different formats and builds.</p>
<p>In this vignette, we’ll use GWAS catalog to obtain an example dataset. For this vignette we’ll use a Breast cancer (BRCA) dataset from <a href="https://www.nature.com/articles/nature24284" class="external-link">Michailidou et al., 2017</a>, which is one that we used in our manuscript. This GWAS was performed on 119078 controls and 137045 cases of Breast cancer.</p>
<p>It’s a big file, and may take a while to download, so here we will show the command to download, but actually cheat and load a subset of data already loaded.</p>
<p>Note that in this particular case we chose a study that contained a single dataset. In other cases there may be more than one. In that case <code>gwascat.download</code> will prompt you with the list of available options, prompting their accession numbers, and asking you to choose a file by its number in the list, if running interactively.</p>
<p>We use it’s PubMed ID (29059683). When running <code>gwascat.download</code> without any other arguments, it will try to get the harmonised files associated with the ID. Harmonised files have been processed by GWAS catalog and are formatted and aligned to the lastest genome build. See <a href="https://www.ebi.ac.uk/gwas/docs/methods/summary-statistics" class="external-link">here</a> for more information.</p>
<p>Once we download the file, we’ll need to prepare it for RápidoPGS. That will involve renaming columns to something RápidoPGS can understand, and this is easy to do with data.table. Also, RápidoPGS use only autosomes, so remove X or Y chromosomes at this step.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gwascat.download.html">gwascat.download</a></span><span class="op">(</span><span class="fl">29059683</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Select the harmonised hg38 file </span></span>
<span><span class="co"># This is equivalent to:</span></span>
<span><span class="co"># ds &lt;- fread("ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST004001-GCST005000/GCST004988/harmonised/29059683-GCST004988-EFO_0000305.h.tsv.gz")</span></span>
<span></span>
<span><span class="co"># Then apply some reformatting</span></span>
<span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/setattr.html" class="external-link">setnames</a></span><span class="op">(</span><span class="va">ds</span>, old <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hm_rsid"</span>,<span class="st">"hm_chrom"</span>,<span class="st">"hm_pos"</span>, <span class="st">"hm_other_allele"</span>, <span class="st">"hm_effect_allele"</span>, <span class="st">"hm_effect_allele_frequency"</span>, <span class="st">"hm_beta"</span>, <span class="st">"standard_error"</span>, <span class="st">"p_value"</span><span class="op">)</span>, new <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SNPID"</span>,<span class="st">"CHR"</span>, <span class="st">"BP"</span>, <span class="st">"REF"</span>,<span class="st">"ALT"</span>,<span class="st">"ALT_FREQ"</span>, <span class="st">"BETA"</span>, <span class="st">"SE"</span>, <span class="st">"P"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ds</span> <span class="op">&lt;-</span> <span class="va">ds</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">SNPID</span>, <span class="va">CHR</span>, <span class="va">BP</span>, <span class="va">REF</span>, <span class="va">ALT</span>, <span class="va">ALT_FREQ</span>, <span class="va">BETA</span>, <span class="va">SE</span>, <span class="va">P</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">ds</span> <span class="op">&lt;-</span> <span class="va">ds</span><span class="op">[</span><span class="va">CHR</span> <span class="op">!=</span><span class="st">"X"</span><span class="op">]</span></span>
<span><span class="va">ds</span><span class="op">$</span><span class="va">CHR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">ds</span><span class="op">$</span><span class="va">CHR</span><span class="op">)</span></span>
<span><span class="va">ds</span> <span class="op">&lt;-</span> <span class="va">ds</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">CHR</span>, <span class="va">BP</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">ds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/na.fail.html" class="external-link">na.omit</a></span><span class="op">(</span><span class="va">ds</span>, cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"BETA"</span>, <span class="st">"ALT_FREQ"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>For illustrative purposes, I took a random subset from this file including 100,000 SNPs from this file, which we’ll use in this tutorial. Bear in mind that the final PGS won’t be a <em>valid</em> model since we randomly removed most SNPs. It will serve for our teaching purposes, though! ;) We can load this file straight away.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ds</span> <span class="op">&lt;-</span> <span class="va">michailidou38</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="applying-quality-control-to-the-dataset">Applying quality control to the dataset<a class="anchor" aria-label="anchor" href="#applying-quality-control-to-the-dataset"></a>
</h2>
<p>The first thing to do is to check what our file looks like. RápidoPGS can handle NAs in crucial columns, so don’t worry too much about that (unless you have all NA for crucial columns, of course!).</p>
<p>A note of caution when dealing with GWAS catalog files, though: since they use a semi-automated pipeline, it is possible that even some columns are present, they are <em>empty</em>, so be careful with that!</p>
<p>Also, RápidoPGS requires allele frequencies, so it’s possible that you need to compute it from an external reference panel (eg. 1000 Genomes Phase III). We don’t cover that step in this vignette, but we might write instructions explaining how to do it in the future.</p>
<p>Lastly, we applied a number of QC steps in our paper, which we won’t apply here, but encourage you to try when using real data. The details of this procedure are described in the paper. You can also take a look at the code <a href="https://github.com/GRealesM/RapidoPGS_paper/blob/master/Preparing_datasets_qcfilt_clean_20210317.R" class="external-link">here</a>.</p>
<p>Let’s now look at the file.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">ds</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     SNPID                CHR               BP                REF           </span></span>
<span><span class="co">##  Length:100000      Min.   : 1.000   Min.   :    19484   Length:100000     </span></span>
<span><span class="co">##  Class :character   1st Qu.: 4.000   1st Qu.: 32314220   Class :character  </span></span>
<span><span class="co">##  Mode  :character   Median : 7.000   Median : 69249390   Mode  :character  </span></span>
<span><span class="co">##                     Mean   : 8.612   Mean   : 78711220                     </span></span>
<span><span class="co">##                     3rd Qu.:13.000   3rd Qu.:114464115                     </span></span>
<span><span class="co">##                     Max.   :22.000   Max.   :248922216                     </span></span>
<span><span class="co">##      ALT               ALT_FREQ           BETA                  SE         </span></span>
<span><span class="co">##  Length:100000      Min.   :0.0051   Min.   :-2.1291000   Min.   :0.00620  </span></span>
<span><span class="co">##  Class :character   1st Qu.:0.0205   1st Qu.:-0.0100000   1st Qu.:0.00740  </span></span>
<span><span class="co">##  Mode  :character   Median :0.1037   Median :-0.0003000   Median :0.01140  </span></span>
<span><span class="co">##                     Mean   :0.2245   Mean   :-0.0009533   Mean   :0.01949  </span></span>
<span><span class="co">##                     3rd Qu.:0.3571   3rd Qu.: 0.0094000   3rd Qu.:0.02540  </span></span>
<span><span class="co">##                     Max.   :0.9949   Max.   : 0.6442000   Max.   :0.83960  </span></span>
<span><span class="co">##        P                N         </span></span>
<span><span class="co">##  Min.   :0.0000   Min.   :256123  </span></span>
<span><span class="co">##  1st Qu.:0.1922   1st Qu.:256123  </span></span>
<span><span class="co">##  Median :0.4470   Median :256123  </span></span>
<span><span class="co">##  Mean   :0.4597   Mean   :256123  </span></span>
<span><span class="co">##  3rd Qu.:0.7174   3rd Qu.:256123  </span></span>
<span><span class="co">##  Max.   :0.9991   Max.   :256123</span></span></code></pre>
<p>In this case, we don’t have any missing data, which is fantastic.</p>
</div>
<div class="section level2">
<h2 id="computing-pgs-using-rápidopgs-single">Computing PGS using RápidoPGS-single<a class="anchor" aria-label="anchor" href="#computing-pgs-using-r%C3%A1pidopgs-single"></a>
</h2>
<p>Let’s now compute our PGS! The build of this example file is <strong>hg38</strong>, so we must tell RápidoPGS about that (default is hg19).</p>
<p>In this new version, we don’t need sample size for case-control datasets. Note that if this was a quantitative trait dataset, we should inform total number of individuals (N parameter). Also, if our dataset had columns reporting the number of individuals for each SNP, we can replace N by a string specifying the column (eg. N = “sample_size”). By doing so, RápidoPGS-single will extract this information directly from the columns.</p>
<p>Let’s get our hands dirty! Let’s compute first a <em>full</em> RápidoPGS-single model.</p>
<p><strong>Advanced use note</strong>: You may want to filter by and align your dataset to an external reference. You can do that with RápidoPGS, too, by specifying the path of your reference file at <code>reference</code> argument. This reference file should have five columns (CHR, BP, REF, ALT, and SNPID) and be in the <strong>same build</strong> as our summary statistics dataset. RápidoPGS currently supports <strong>hg19</strong> and <strong>hg38</strong> builds.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">full_PGS</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rapidopgs_single.html">rapidopgs_single</a></span><span class="op">(</span><span class="va">ds</span>, trait <span class="op">=</span> <span class="st">"cc"</span>, build <span class="op">=</span> <span class="st">"hg38"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Assigning LD blocks...</span></span></code></pre>
<pre><code><span><span class="co">## Done!</span></span></code></pre>
<pre><code><span><span class="co">## Computing a RapidoPGS-single model for a case-control dataset...</span></span></code></pre>
<p>Well, that was <em>rápido</em>! With increasingly big datasets, it will take a bit longer, of course.</p>
<p>Let’s take a look.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">full_PGS</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          SNPID CHR     BP REF ALT ALT_FREQ    BETA     SE      P      N</span></span>
<span><span class="co">## 1: rs371431021   1 629750   G   A   0.0145  0.0002 0.0442 0.9969 256123</span></span>
<span><span class="co">## 2: rs577189614   1 633285   G   A   0.0115  0.0201 0.0705 0.7754 256123</span></span>
<span><span class="co">## 3: rs142849724   1 847691   C   T   0.0191  0.0490 0.0353 0.1648 256123</span></span>
<span><span class="co">## 4: rs143609865   1 864813   A   T   0.0382 -0.0049 0.0227 0.8276 256123</span></span>
<span><span class="co">## 5: rs372425838   1 866967   C   T   0.0715  0.0013 0.0190 0.9474 256123</span></span>
<span><span class="co">## 6: rs138192888   1 937087   C   G   0.0061 -0.0471 0.0593 0.4276 256123</span></span>
<span><span class="co">##    ld.block          ppi        weight</span></span>
<span><span class="co">## 1:        1 2.153954e-05  4.307908e-09</span></span>
<span><span class="co">## 2:        1 3.440498e-05  6.915402e-07</span></span>
<span><span class="co">## 3:        1 4.416140e-05  2.163908e-06</span></span>
<span><span class="co">## 4:        1 1.151862e-05 -5.644124e-08</span></span>
<span><span class="co">## 5:        1 9.461824e-06  1.230037e-08</span></span>
<span><span class="co">## 6:        1 3.791770e-05 -1.785924e-06</span></span></code></pre>
<p>We see three new columns: <strong>ld.block</strong> corresponds to the <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4731402/" class="external-link">ld-detect</a> LD block number assigned to the SNP (see manuscript for more details of where this comes from), <strong>ppi</strong> is the posterior probability that the SNP is causal, and <strong>weight</strong> is the weighted effect size (BETA * ppi) - and the column we’re interested in.</p>
<div class="section level3">
<h3 id="applying-different-thresholds">Applying different thresholds<a class="anchor" aria-label="anchor" href="#applying-different-thresholds"></a>
</h3>
<p>Imagine that we want a small portable PGS. We could use a threshold (eg. keep only SNPs with ppi larger than <span class="math inline">\(1e^{-4}\)</span>, a reasonable threshold) or keep the top SNPs with largest weights (in either way).</p>
<p>We can do that using RápidoPGS, let’s see how by using a <span class="math inline">\(1e^{-4}\)</span> threshold.</p>
<p>For accuracy reasons, we recommend recomputing the PGS on just these SNPs after the threshold was applied, so <code>recalc = TRUE</code> by default.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">PGS_1e4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rapidopgs_single.html">rapidopgs_single</a></span><span class="op">(</span><span class="va">ds</span>, trait <span class="op">=</span><span class="st">"cc"</span>, build <span class="op">=</span> <span class="st">"hg38"</span>, filt_threshold <span class="op">=</span> <span class="fl">1e-4</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Assigning LD blocks...</span></span></code></pre>
<pre><code><span><span class="co">## Done!</span></span></code></pre>
<pre><code><span><span class="co">## Computing a RapidoPGS-single model for a case-control dataset...</span></span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">PGS_1e4</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          SNPID CHR      BP REF ALT ALT_FREQ    BETA     SE        P      N</span></span>
<span><span class="co">## 1: rs549049947   1 1016953   A   G   0.0051 -0.1720 0.1069 0.107700 256123</span></span>
<span><span class="co">## 2:   rs4452985   1 1023646   G   A   0.0066 -0.1051 0.0583 0.071380 256123</span></span>
<span><span class="co">## 3:  rs74346965   1 2435573   G   A   0.0150  0.0803 0.0425 0.059090 256123</span></span>
<span><span class="co">## 4: rs111235033   1 3030855   G   A   0.0089  0.1161 0.0445 0.009074 256123</span></span>
<span><span class="co">## 5:   rs1971331   1 3256718   A   G   0.9482 -0.0501 0.0179 0.005214 256123</span></span>
<span><span class="co">## 6:  rs79983520   1 3540815   G   A   0.0068  0.1047 0.0597 0.079340 256123</span></span>
<span><span class="co">##    ld.block          ppi        weight</span></span>
<span><span class="co">## 1:        1 0.0001289612 -2.218132e-05</span></span>
<span><span class="co">## 2:        1 0.0001250847 -1.314640e-05</span></span>
<span><span class="co">## 3:        1 0.0001146317  9.204928e-06</span></span>
<span><span class="co">## 4:        2 0.0005553243  6.447315e-05</span></span>
<span><span class="co">## 5:        2 0.0004335087 -2.171879e-05</span></span>
<span><span class="co">## 6:        2 0.0001172193  1.227286e-05</span></span></code></pre>
<p>You can omit recalculation by setting that argument to <code>FALSE</code></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">PGS_1e4_norecalc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rapidopgs_single.html">rapidopgs_single</a></span><span class="op">(</span><span class="va">ds</span>, trait <span class="op">=</span><span class="st">"cc"</span>, build <span class="op">=</span> <span class="st">"hg38"</span>, filt_threshold <span class="op">=</span> <span class="fl">1e-4</span>, recalc <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Assigning LD blocks...</span></span></code></pre>
<pre><code><span><span class="co">## Done!</span></span></code></pre>
<pre><code><span><span class="co">## Computing a RapidoPGS-single model for a case-control dataset...</span></span></code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">PGS_1e4_norecalc</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          SNPID CHR      BP REF ALT ALT_FREQ    BETA     SE        P      N</span></span>
<span><span class="co">## 1: rs549049947   1 1016953   A   G   0.0051 -0.1720 0.1069 0.107700 256123</span></span>
<span><span class="co">## 2:   rs4452985   1 1023646   G   A   0.0066 -0.1051 0.0583 0.071380 256123</span></span>
<span><span class="co">## 3:  rs74346965   1 2435573   G   A   0.0150  0.0803 0.0425 0.059090 256123</span></span>
<span><span class="co">## 4: rs111235033   1 3030855   G   A   0.0089  0.1161 0.0445 0.009074 256123</span></span>
<span><span class="co">## 5:   rs1971331   1 3256718   A   G   0.9482 -0.0501 0.0179 0.005214 256123</span></span>
<span><span class="co">## 6:  rs79983520   1 3540815   G   A   0.0068  0.1047 0.0597 0.079340 256123</span></span>
<span><span class="co">##    ld.block          ppi        weight</span></span>
<span><span class="co">## 1:        1 0.0001287698 -2.214840e-05</span></span>
<span><span class="co">## 2:        1 0.0001248991 -1.312689e-05</span></span>
<span><span class="co">## 3:        1 0.0001144616  9.191265e-06</span></span>
<span><span class="co">## 4:        2 0.0005543590  6.436108e-05</span></span>
<span><span class="co">## 5:        2 0.0004327552 -2.168103e-05</span></span>
<span><span class="co">## 6:        2 0.0001170155  1.225152e-05</span></span></code></pre>
<p>And what if we want the <em>top 10</em> SNPs? Just change the <code>filt_threshold</code> argument. If it’s larger than 1, RápidoPGS will understand that you want a top list, rather than a thresholded result.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">PGS_top10</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rapidopgs_single.html">rapidopgs_single</a></span><span class="op">(</span><span class="va">ds</span>, trait <span class="op">=</span><span class="st">"cc"</span>, build <span class="op">=</span> <span class="st">"hg38"</span>, filt_threshold <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Assigning LD blocks...</span></span></code></pre>
<pre><code><span><span class="co">## Done!</span></span></code></pre>
<pre><code><span><span class="co">## Computing a RapidoPGS-single model for a case-control dataset...</span></span></code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">PGS_top10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          SNPID CHR        BP REF ALT ALT_FREQ    BETA     SE         P      N</span></span>
<span><span class="co">## 1:  rs72812748  10  22146738   T   C   0.0189  0.1946 0.0223 2.288e-18 256123</span></span>
<span><span class="co">## 2:  rs73122135   5  56971481   G   A   0.0488  0.1939 0.0139 2.397e-44 256123</span></span>
<span><span class="co">## 3:  rs78238929   2 216858846   G   A   0.0131 -0.2070 0.0310 2.552e-11 256123</span></span>
<span><span class="co">## 4: rs142053364  17  42545533   A   C   0.0111  0.1627 0.0300 5.871e-08 256123</span></span>
<span><span class="co">## 5:   rs6912323   6 151637477   G   T   0.0700  0.1814 0.0117 4.640e-54 256123</span></span>
<span><span class="co">## 6:  rs73171018  22  40665687   C   A   0.0410  0.1350 0.0159 1.815e-17 256123</span></span>
<span><span class="co">##    ld.block       ppi     weight</span></span>
<span><span class="co">## 1:      818 1.0000000  0.1946000</span></span>
<span><span class="co">## 2:      442 1.0000000  0.1939000</span></span>
<span><span class="co">## 3:      204 0.9999771 -0.2069953</span></span>
<span><span class="co">## 4:     1203 0.9631757  0.1567087</span></span>
<span><span class="co">## 5:      577 1.0000000  0.1814000</span></span>
<span><span class="co">## 6:     1355 1.0000000  0.1350000</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>So those are examples of basic <em>RápidoPGS-single</em> usage!</p>
<p>Drop us a line if you encounter problems, and we’ll be happy to help.</p>
<p>Good luck!</p>
<p>Guillermo</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Guillermo Reales, Chris Wallace.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
