<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Compute PGS from GWAS summary statistics using Bayesian sum of single-effect (SuSiE) linear regression using z scores — rapidopgs_multi • RápidoPGS</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Compute PGS from GWAS summary statistics using Bayesian sum of single-effect (SuSiE) linear regression using z scores — rapidopgs_multi"><meta property="og:description" content="'rapidopgs_multi computes PGS from a from GWAS summary statistics
using Bayesian sum of single-effect (SuSiE) linear regression using z scores"><meta property="og:image" content="http://GRealesM.github.io/RapidoPGS/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RápidoPGS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.3.0.9002</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Computing_RapidoPGSmulti.html">Computing a PGS using RápidoPGS-multi</a>
    </li>
    <li>
      <a href="../articles/Computing_RapidoPGSsingle.html">Computing a PGS using RápidoPGS-single and GWAS catalog</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/GRealesM/RapidoPGS" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Compute PGS from GWAS summary statistics using Bayesian sum of single-effect (SuSiE) linear regression using z scores</h1>
    <small class="dont-index">Source: <a href="https://github.com/GRealesM/RapidoPGS/R/rapfunc.R" class="external-link"><code>R/rapfunc.R</code></a></small>
    <div class="hidden name"><code>rapidopgs_multi.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>'<code>rapidopgs_multi</code> computes PGS from a from GWAS summary statistics
using Bayesian sum of single-effect (SuSiE) linear regression using z scores</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">rapidopgs_multi</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  reference <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  LDmatrices <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  N <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  build <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hg19"</span>, <span class="st">"hg38"</span><span class="op">)</span>,</span>
<span>  trait <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cc"</span>, <span class="st">"quant"</span><span class="op">)</span>,</span>
<span>  ncores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  alpha.block <span class="op">=</span> <span class="fl">1e-04</span>,</span>
<span>  alpha.snp <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  sd.prior <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ancestry <span class="op">=</span> <span class="st">"EUR"</span>,</span>
<span>  LDblocks <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>a data.table containing GWAS summary statistic dataset
with all required information.</p></dd>


<dt id="arg-reference">reference<a class="anchor" aria-label="anchor" href="#arg-reference"></a></dt>
<dd><p>a string representing the path to the directory containing
the reference panel (eg. "../ref-data").</p></dd>


<dt id="arg-ldmatrices">LDmatrices<a class="anchor" aria-label="anchor" href="#arg-ldmatrices"></a></dt>
<dd><p>a string representing the path to the directory containing
the pre-computed LD matrices.</p></dd>


<dt id="arg-n">N<a class="anchor" aria-label="anchor" href="#arg-n"></a></dt>
<dd><p>a numeric indicating the number of individuals used to generate input
GWAS dataset, or a string indicating the column name containing per-SNP sample size.</p></dd>


<dt id="arg-build">build<a class="anchor" aria-label="anchor" href="#arg-build"></a></dt>
<dd><p>a string indicating the genome build. 'hg19' and 'hg38' are supported. Note that your LD matrices or reference panel should match the build.</p></dd>


<dt id="arg-trait">trait<a class="anchor" aria-label="anchor" href="#arg-trait"></a></dt>
<dd><p>a string indicating if trait is a case-control ("cc") or quantitative ("quant").</p></dd>


<dt id="arg-ncores">ncores<a class="anchor" aria-label="anchor" href="#arg-ncores"></a></dt>
<dd><p>a numeric specifying the number of cores (CPUs) to be used.
If using pre-computed LD matrices, one core is enough for best performance.</p></dd>


<dt id="arg-alpha-block">alpha.block<a class="anchor" aria-label="anchor" href="#arg-alpha-block"></a></dt>
<dd><p>a numeric threshold for minimum P-value in LD blocks.
Blocks with minimum P above <code>alpha.block</code> will be skipped. Default: 1e-4.</p></dd>


<dt id="arg-alpha-snp">alpha.snp<a class="anchor" aria-label="anchor" href="#arg-alpha-snp"></a></dt>
<dd><p>a numeric threshold for P-value pruning within LD block.
SNPs with P above <code>alpha.snp</code> will be removed. Default: 0.01.</p></dd>


<dt id="arg-sd-prior">sd.prior<a class="anchor" aria-label="anchor" href="#arg-sd-prior"></a></dt>
<dd><p>the prior specifies that BETA at causal SNPs
follows a centred normal distribution with standard deviation
sd.prior.
If NULL (default) it will be automatically estimated (recommended).</p></dd>


<dt id="arg-ancestry">ancestry<a class="anchor" aria-label="anchor" href="#arg-ancestry"></a></dt>
<dd><p>a string indicating the ancestral population (DEFAULT: "EUR", European).
If using an alternative population, bear in mind that your LD matrices or reference must be from the same population.
You'll also need to provide matching LD.blocks via the LDblocks argument.</p></dd>


<dt id="arg-ldblocks">LDblocks<a class="anchor" aria-label="anchor" href="#arg-ldblocks"></a></dt>
<dd><p>a string indicating the path to an alternative LD block file in .RData format. Only required for non-European PGS.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>a data.table containing the sumstats dataset with
computed PGS weights.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This function will take a GWAS summary statistic dataset as an input,
will assign LD blocks to it, then use user-provided LD matrices or a preset
reference panel in Plink format to compute LD matrices for each block.
Then SuSiE method will be used to compute posterior probabilities of variants to be causal
and generate PGS weights by multiplying those posteriors by effect sizes (\(\beta\)).
Unlike <code>rapidopgs_single</code>, this approach will assume one or more causal variants.</p>
<p>The GWAS summary statistics file to compute PGS using our method must contain
the following minimum columns, with these exact column names:</p><dl><dt>CHR</dt>
<dd><p>Chromosome</p></dd>

<dt>BP</dt>
<dd><p>Base position (in GRCh37/hg19).</p></dd>

<dt>REF</dt>
<dd><p>Reference, or non-effect allele</p></dd>

<dt>ALT</dt>
<dd><p>Alternative, or effect allele, the one \(\beta\) refers to</p></dd>

<dt>BETA</dt>
<dd><p>\(\beta\) (or log(OR)), or effect sizes</p></dd>

<dt>SE</dt>
<dd><p>standard error of \(\beta\)</p></dd>

<dt>P</dt>
<dd><p>P-value for the association test</p></dd>


</dl><p>In addition, quantitative traits must have the following extra column:</p><dl><dt>ALT_FREQ</dt>
<dd><p>Minor allele frequency.</p></dd>


</dl><p>Also, for quantitative traits, sample size must be supplied, either as a number,
or indicating the column name, for per-SNP sample size datasets (see below).
Other columns are allowed, and will be ignored.</p>
<p>Reference panel should be divided by chromosome, in Plink format.
Both reference panel and summary statistic dataset should be in GRCh37/hg19.
For 1000 Genomes panel, you can use <code>create_1000G</code> function to set it up
automatically.</p>
<p>If prefer to use LD matrices, you must indicate the path to the directory
where they are stored. They must be in RDS format, named LD_chrZ.rds (where
Z is the 1-22 chromosome number). If you don't have LD matrices already,
we recommend downloading those gently provided by Prive et al., at
<a href="https://figshare.com/articles/dataset/European_LD_reference/13034123" class="external-link">https://figshare.com/articles/dataset/European_LD_reference/13034123</a>.
These matrices were computed using for 1,054,330 HapMap3 variants based on
362,320 European individuals of the UK biobank.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Guillermo Reales, Chris Wallace</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="va">ss</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="op">(</span></span></span>
<span class="r-in"><span>    CHR<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">20</span>,<span class="fl">14</span>,<span class="fl">2</span>,<span class="fl">4</span>,<span class="fl">6</span>,<span class="fl">6</span>,<span class="fl">21</span>,<span class="fl">13</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>    BP<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1479959</span>, <span class="fl">13000913</span>, <span class="fl">29107209</span>, <span class="fl">203573414</span>, <span class="fl">57331393</span>, <span class="fl">11003529</span>, <span class="fl">149256398</span>, </span></span>
<span class="r-in"><span>        <span class="fl">25630085</span>, <span class="fl">79166661</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>    REF<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"C"</span>,<span class="st">"C"</span>,<span class="st">"C"</span>,<span class="st">"T"</span>,<span class="st">"G"</span>,<span class="st">"C"</span>,<span class="st">"C"</span>,<span class="st">"G"</span>,<span class="st">"T"</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>    ALT<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>,<span class="st">"T"</span>,<span class="st">"T"</span>,<span class="st">"A"</span>,<span class="st">"A"</span>,<span class="st">"A"</span>,<span class="st">"T"</span>,<span class="st">"A"</span>,<span class="st">"C"</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>    BETA<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.012</span>,<span class="fl">0.0079</span>,<span class="fl">0.0224</span>,<span class="fl">0.0033</span>,<span class="fl">0.0153</span>,<span class="fl">0.058</span>,<span class="fl">0.0742</span>,<span class="fl">0.001</span>,<span class="op">-</span><span class="fl">0.0131</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    SE<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.0099</span>,<span class="fl">0.0066</span>,<span class="fl">0.0203</span>,<span class="fl">0.0171</span>,<span class="fl">0.0063</span>,<span class="fl">0.0255</span>,<span class="fl">0.043</span>,<span class="fl">0.0188</span>,<span class="fl">0.0074</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    P<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.2237</span>,<span class="fl">0.2316</span>,<span class="fl">0.2682</span>,<span class="fl">0.8477</span>,<span class="fl">0.01473</span>,<span class="fl">0.02298</span>,<span class="fl">0.08472</span>,<span class="fl">0.9573</span>,<span class="fl">0.07535</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">PGS</span> <span class="op">&lt;-</span> <span class="fu">rapidopgs_multi</span><span class="op">(</span><span class="va">ss</span>, reference <span class="op">=</span> <span class="st">"ref-data/"</span>, N <span class="op">=</span> <span class="fl">20000</span>, build <span class="op">=</span> <span class="st">"hg19"</span>, trait<span class="op">=</span><span class="st">"cc"</span>, ncores<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Guillermo Reales, Chris Wallace.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer></div>






  </body></html>

